### ISSUE TITLE
standard library feature request: API for outputting progress to the terminal

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2865]
Sometimes Zig must spend a long time performing computations before responding:

 * When a compilation is taking a long time.
 * When cross compiling musl, glibc, or mingw for the first time, and the respective libc must be built from source.

Usually, Zig should output nothing. In general, if everything is going as expected, nothing should be displayed. However, if an operation takes longer than a certain amount of time, it becomes ambiguous whether something is hanging, or it is going as expected. In such a case, progress should be displayed.

I would like an API something like this:

```zig
var my_progress_node = ProgressNode.start(optional_parent_progress_node, optional_estimated_count_of_items_to_do);
```

Later when some sense of progress has been made:

```zig
my_progress_node.update(optional_string_what_is_it_doing_now, optional_how_many_are_done);
```

And finally when the item is complete:

```zig
my_progress_node.end();
```

My idea for this is that if the operation takes less than 200 milliseconds, nothing is displayed. If the operation takes longer than this, then the implementation can decide how to display the progress.

The API could have configuration options for how the output should be displayed:
 * (Default) - detect if stderr is a terminal, and if so, use terminal escape codes to display nice progress
 * Unconditionally silenced (probably `--color off` would do this)
 * (future idea, yagni) expose the progress over the Language Server Protocol or other system

This can be coded in Zig rather than C++ because the feature does not depend on itself. It can be exposed to the C++ code via libuserland, just like `zig fmt` and `zig translate-c-2`.

**Comment by shritesh** [comment]
It’d be cool if sending SIGINFO would show the current progress. `dd` does this.

**Comment by andrewrk** [comment]
Landed in e42d86b657c2fae093fb2545e8b4b85614a0c906
